enterprise/backend/src/metabase_enterprise/advanced_permissions/query_processor/middleware/permissions.clj

```diff
@@ -31,13 +31,14 @@
   [{db-id :database, :as query}]
   ;; Remove the :native key (containing the transpiled MBQL) so that this helper function doesn't think the query is
   ;; a native query. Actual native queries are dispatched to a different method by the :type key.
-  (let [table-ids   (query-perms/query->source-table-ids (dissoc query :native))
-        table-perms (into #{}
-                          (map (fn [table-id]
-                                 (if (= table-id ::query-perms/native)
-                                   (data-perms/native-download-permission-for-user api/*current-user-id* db-id)
-                                   (data-perms/table-permission-for-user api/*current-user-id* :perms/download-results db-id table-id)))
-                               table-ids))]
+  (let [{:keys [table-ids card-ids native?]} (query-perms/query->source-ids (dissoc query :native))
+        table-perms (if (or native? (seq card-ids))
+                      ;; If we detect any native subqueries/joins, even with source-card IDs, require full native
+                      ;; download perms
+                      #{(data-perms/native-download-permission-for-user api/*current-user-id* db-id)}
```

Comment:
So `:qp/stage-is-from-source-card` implies native? I.e. doesn't get set for source cards with mbql queries?